### PR TITLE
Handle union types for typed properties and docblocks

### PR DESCRIPTION
**Pull request by dgafka**
This handle union types for typed properties and DocBlock type resolver. 

The solution allows for defining union types for above, which was previously failing during cache building phase. 
As union type is non implicit in what we are actually serializing / deserializing, we are not trying to recognize which type user want to use in specific context.

How does serialization works?
Serializing knows which type to use, as it actually holds the value that is going to be serialized. It use that value to infer the type and do the serialization. 

How does deserialization works? 
The deserialization will throw an exception as we can't guess, which type user would like to pick. 
If developer will define Type Attribute, it will be chosen as default one. 
This is a place for improvement. Possible we could introduce custom union deserializing handler, that would decide on his own what class he would like to deserialize too. 

| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | yes
| Doc updated   | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | https://github.com/schmittjoh/serializer/issues/1282 https://github.com/schmittjoh/serializer/issues/1302
| License       | MIT

### CODE REVIEW

**Comment by dgafka** [comment]
ping @yyaremenko @gam6itko :)

**Comment by goetas** [comment]
i'm sorry for the delay, @dgafka  could you please rebase this?

**Comment by dgafka** [comment]
@goetas branch is up to date with master now :) 
I am not sure, how to fix phpstan or phpcs for lower PHP versions, can you do it?

**Comment by davidwindell** [comment]
Just tested this on our app and can confirm it seems to work well for us.

**Comment by dgafka** [comment]
Hey @goetas,
Can this be merged? :)

**Review by scyzoryck — CHANGES_REQUESTED** [review]
@dgafka  - Could you take a look at the code style issues that has been reported by the coding standard, please? I believe they can be fixed automatically by the tool. Also you need to exclude your fixture from PHPStan analysis - as in CI it is running with PHP7.4 that do not support union types.

**Comment by dgafka** [comment]
@scyzoryck how to fix that:

```
  Line   tests/Metadata/Driver/TypedPropertiesDriverTest.php                                                                                                                            
 ------ ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 
  15     ParseError (syntax error, unexpected '|', expecting variable (T_VARIABLE)) thrown while looking for class JMS\Serializer\Tests\Fixtures\TypedProperties\UnionTypedProperties.  
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                                                                                                             
  53     ParseError (syntax error, unexpected '|', expecting variable (T_VARIABLE)) thrown while looking for class JMS\Serializer\Tests\Fixtures\TypedProperties\UnionTypedProperties.  
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols 
```

as excluding it from phpstan analyse does not change a thing:

```
    excludePaths:
        - %currentWorkingDirectory%/src/Annotation/ReadOnly.php
        - %currentWorkingDirectory%/tests/Fixtures/TypedProperties/UnionTypedProperties.php
```

**Comment by dgafka** [comment]
I've put the test under phpstan ignore too. Tests are passing now.

File: tests/Metadata/Driver/UnionTypedPropertiesDriverTest.php [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+use Metadata\ClassMetadata;
+use PHPUnit\Framework\TestCase;
+use ReflectionClass;
+
+final class UnionTypedPropertiesDriverTest extends TestCase
+{
+    protected function setUp(): void
+    {
+        if (PHP_VERSION_ID < 80000) {
+            $this->markTestSkipped(sprintf('%s requires PHP 7.4', TypedPropertiesDriver::class));
```
Comment (scyzoryck) [inline tests/Metadata/Driver/UnionTypedPropertiesDriverTest.php:21]:
```suggestion
            $this->markTestSkipped(sprintf('%s requires PHP 8.0', TypedPropertiesDriver::class));
```

File: phpstan.neon.dist [hunk 2]
```diff
@@ -12,4 +12,8 @@ parameters:
         - %currentWorkingDirectory%/tests
     excludePaths:
         - %currentWorkingDirectory%/src/Annotation/ReadOnly.php
-     
+        - %currentWorkingDirectory%/tests/Fixtures/TypedProperties/UnionTypedProperties.php
+        - %currentWorkingDirectory%/tests/Fixtures/TypedProperties/UnionTypedProperties.php
+        - %currentWorkingDirectory%/tests/Metadata/Driver/UnionTypedPropertiesDriverTest.php
+        - %currentWorkingDirectory%/tests/Serializer/BaseSerializationTest.php
```
Comment (scyzoryck) [inline phpstan.neon.dist:18]: I will check tomorrow if we can setup differently - it would be better to do not ignore test files.

Comment (scyzoryck) [reply phpstan.neon.dist:18]: I didn't find any solution to exclude it. May we can split the tests for PHP8.0 only into separate class and ignore them for PHP7.4 static analytics? Configuration for the jobs can be found here: `.github/workflows/static-analysis.yaml`.

**Comment by dgafka** [comment]
Extracted separate phpstan analysis for PHP 7.4.
The rest of the versions are running against those files :)

**Review by scyzoryck — APPROVED** [review]
Thanks! Looks good for me! @goetas - Could you take a look, please?

**Pull request merged**